### ISSUE TITLE
bug: Structure referencing itself

### ISSUE THREAD

**Issue by gaetbout** [starkware-libs/cairo#2249]
# Bug Report

**Cairo version:**

Tested at: eaa11531ec77a4f725a7f16c7562c16024ee8487 

**Current behavior:**

Trying to run this code:
```
// Core lib imports
use option::OptionTrait;
use box::Box;

#[derive(Drop)]
struct Node {
    value: felt,
    left: Option::<Box::<Node>>,
}

// impl NodeDrop of Drop::<Node>;
impl BoxNodeDrop of Drop::<Box::<Node>>;
impl OptionBoxNodeDrop of Drop::<Option::<Box::<Node>>>;

#[test]
#[available_gas(2000000)]
fn simple_test() {
    let bst = Node {
            value :12, 
            left: Option::None(()),
        };
        
    assert(bst.value == 12, 'value should be 12');
    // assert(bst.left.is_none(), 'left should be none');

}
```
I run into: 
`thread 'main' panicked at 'Internal error, cycle detected:`
And if I uncomment the last assertion, I run into:
`error: Variable was previously moved.`

**Expected behavior:**

It should be possible to create such structure.

**Related commits:**
3ef87ab6ad81cfb46f78ecf7cfcc9a876ca82920
7400dc52c65e6ee3eb52bd786be00ef2a61d0575

**Comment by spapinistarkware** [comment]
The Variabled moved happens because You didnt implement Copy on all the types. Another solution would be to destructure bst to its fields.
I'm looking into the main issue though.

**Comment by gaetbout** [comment]
I have a local version that:
```
impl Drop::box::node
impl Drop::Option::box::node
impl Copy::box::node
impl Copy::Option::box::node
```
And I still have Cycle detected :(

**Comment by dubzn** [comment]
+1 having same issue, even building from main branch

**Comment by spapinistarkware** [comment]
Most of the work to enable this was done. The remaining task is to fix the interning in sierra-generator. It's currently in hold, but if someone is brave enough to take this, i can guide him.

**Comment by orizi** [comment]
Fixed now on 2.1.1 and main.
